ament_mypy/ament_mypy/main.py

```diff
@@ -0,0 +1,257 @@
+            file=sys.stderr
+        )
+        return 1
+
+    normal_report, error_messages, exit_code = generate_mypy_report(
+        args.config_file,
+        filenames,
+        args.cache_dir
+    )
+
+    if error_messages:
+        print('mypy error encountered', file=sys.stderr)
+        print(error_messages, file=sys.stderr)
+        return exit_code
+
+    errors_parsed = get_errors(normal_report)
+
+    print('\n{} files checked'.format(len(filenames)))
+    if not normal_report:
+        print('No errors found')
+    else:
+        print('{} errors'.format(len(errors_parsed)))
+
+    print(normal_report)
+
+    print('\nChecked files:')
+    print(''.join(['\n* {}'.format(f) for f in filenames]))
+
+    # generate xunit file
+    if args.xunit_file:
+        folder_name = os.path.basename(os.path.dirname(args.xunit_file))
+        file_name = os.path.basename(args.xunit_file)
+        suffix = '.xml'
+        if file_name.endswith(suffix):
+            file_name = file_name[:-len(suffix)]
+            suffix = '.xunit'
+            if file_name.endswith(suffix):
+                file_name = file_name[:-len(suffix)]
+        testname = '{}.{}'.format(folder_name, file_name)
+
+        xml = get_xunit_content(errors_parsed, testname, filenames, time.time() - start_time)
+        path = os.path.dirname(os.path.abspath(args.xunit_file))
+        if not os.path.exists(path):
+            os.makedirs(path)
+        with open(args.xunit_file, 'w') as f:
+            f.write(xml)
+
+    return exit_code
+
+
+def generate_mypy_report(config_file: str,
+                         paths: List[str],
+                         cache_dir: str) -> Tuple[str, str, int]:
+    mypy_argv = []
+    mypy_argv.append('--cache-dir')
+    mypy_argv.append(str(cache_dir))
+    if cache_dir == os.devnull:
+        mypy_argv.append('--no-incremental')
+    if config_file is not None:
+        mypy_argv.append('--config-file')
+        mypy_argv.append(str(config_file))
+    if not os.environ.get('MYPYPATH') and os.environ.get('PYTHONPATH') and os.environ.get('COLCON_PREFIX_PATH'):
+        # Filter PYTHONPATHs by colcon prefixes
+        python_paths = os.environ['PYTHONPATH'].split(':')
+        colcon_paths = os.environ['COLCON_PREFIX_PATH'].split(':')
+        intersecting_paths = []
+        for python_path in python_paths:
+            for colcon_path in colcon_paths:
+                # The colcon prefix is pointing at the install space, but ideally we could extract
+                # both install-space and build-space paths from PYTHONPATH. There doesn't seem to be
+                # a way to extract these programatically, so we'll assume that the defaults are being
+                # used and the install- and build-space share a parent directory. This assumption will
+                # not be valid if the `--build-base` or `--install-base` options were used for
+                # `colcon build`.
```

Comment:
This seems fairly fragile.

Reply:
@dirk-thomas input would be appreciated regarding how to make it less so. This seems to be the best option so far.

Reply:
> the install- and build-space share a parent directory

The assumption will also be wrong if you use an isolated install (which is the default).

Why do you need the build space in the first place?

Reply:
> The assumption will also be wrong if you use an isolated install (which is the default).

To clarify, the build- and install-spaces discussed here are the "install" and "build" directories used by default, which are used regardless of a merged or isolated install, it's just a question of how deeply things are nested under them, no? The only time that assumption breaks down is if the user specifically puts them elsewhere.

> Why do you need the build space in the first place?

Perhaps it's a side effect of a `--symlink-install`, but:

```
>>> import ros2cli
>>> ros2cli.__file__
'/home/ubuntu/ros2_ws/build/ros2cli/ros2cli/__init__.py'
```

Packages are in there. We can either not follow imports at all, or we can try to follow all ros2-related ones.

Reply:
> Perhaps it's a side effect of a --symlink-install

:+1:

How about determining the actual path with the exact snippet you just used? That will get rid of the assumptions,

Reply:
> How about determining the actual path with the exact snippet you just used?

Wouldn't that require specific knowledge of every package we care about? That doesn't seem feasible to me, the packages being linted can import whatever they want, limited only by the `PYTHONPATH`, which is why we're basing `MYPYPATH` upon it.

Reply:
Indeed that would require knowledge about the provided Python package.

Reply:
Yeah... no such luck :disappointed: . Without an environment variable or other programmatic way to determine the build directory, I'm afraid we're left with the current admittedly-not-ideal-but-hopefully-works-in-most-common-cases method.

Reply:
Maybe I don't understand the issue properly, but why is the `COLCON_PREFIX_PATH` needed at all? Isn't the `PYTHONPATH` sufficient to resolve all the imports? If not wouldn't running the code fail?

Reply:
Hey there @wjwwood! Actually, no-- mypy doesn't use the `PYTHONPATH` for finding imports. See this for some more info:

    https://mypy.readthedocs.io/en/latest/running_mypy.html#finding-imports

Basically, it implements its own variable, `MYPYPATH`, which we're setting here. Without it, it won't find any imports that aren't contained in the package that's being linted. This is viewed as generally okay for system imports as they often don't use static typing anyway, but we thought it would be useful to add ROS 2 dependencies by default since many of them do, and we'd like to see it become more common. The options are outlined in [this comment](https://github.com/ament/ament_lint/pull/154#issuecomment-515119342), but I'll reiterate them here for continuity:

1. Continue trying to support as many ROS 2 dependencies in as many scenarios as possible (current solution), allowing users to ignore ones they don't care about as they see fit
2. Stop trying to handle `--symlink-install`s, and just handle imports contained within `COLCON_PREFIX_PATH` instead of its parent directory.
3. Disable following imports at all, and just check the project being linted and not how it uses its dependencies (i.e. get rid of the use of `MYPYPATH` completely).

All of these options are valid, and we're happy using any of them. We believe (1) provides the most functionality that can be easily pared back by individual packages using it. (2) and (3) are fine as well, although if we remove the functionality here it will be difficult for individual packages and users who want to check their ROS 2 imports to do so.

Reply:
@dirk-thomas explained to me in person. I think this might actually be a mypy bug then. I suggested you guys make an issue and ask them, assuming this test case fails and exposes the root issue:

- make a virtualenv
- install a python project with `pip --editable` (so it's using `.pth` files)
- make a script that imports and uses that python project
- ensure the virtualenv is on the `PYTHONPATH`
- run mypy and see that it fails to find the imports for that python project

If this is the error case you're trying to work around here, then in my opinion that's a bug in mypy because it isn't following `.pth` files which is something the interpreter does do.

Reply:
Reading [their documentation](https://mypy.readthedocs.io/en/latest/running_mypy.html#finding-imports) it sounds like this is by design, not a bug:

> First, mypy has its own search path. This is computed from the following items:
>
>- The MYPYPATH environment variable (a colon-separated list of directories).
>- The mypy_path config file option.
>- The directories containing the sources given on the command line (see below).
>- The installed packages marked as safe for type checking (see PEP 561 support)
>- The relevant directories of the typeshed repo.

Reply:
What about setting the packages as "type safe" in the `setup.py`? Are we doing that for any of the packages you want to check? https://mypy.readthedocs.io/en/latest/installed_packages.html#making-pep-561-compatible-packages

Reply:
I'm not sure if make sense to introduce knowledge about colcon in the first place. AFAIK, there is no reference to colcon in any of the ament packages.

Unless there's another way to PYTHONPATH, I'd suggest we disable following imports for now.

Reply:
> Unless there's another way to filter PYTHONPATH, I'd suggest we disable following imports for now.

Fine by us. We're poking at the PEP 561 spec now, will report back shortly.

Reply:
Yeah, PEP 561-compliant ROS 2 packages don't seem to be picked up. We'll go ahead and remove this logic (i.e. item (3) from [this comment](https://github.com/ament/ament_lint/pull/154/files#r307498685)).